### ISSUE TITLE
NCCL hung  when run Megatron pre-training

### ISSUE THREAD

**Issue by yanminjia** [NVIDIA/nccl#1780]
We implemented kernel-free send/recv based on NCCL code. Basically cuda kernel function will not be launched when send/recv is called by the application such as pytorch/megatron. Therefore the work of memory copy is off loaded to the CPU side. Specifically, in case of P2P communication across 2 nodes via NET transport, on the sender side, the data will be coped to the channel buffer from source address by calling cudaMemcpyAsync on the CPU side firstly.  Likewise, on the receiver side, the received data in the channel buffer is copied to the destination address by calling cudaMemcpyAsync on the CPU side finally. 

the advantage of kernel-free send/recv implementation is that the SM resource could be saved for forward/backward computation without send/recv kernel functions launched In case of Megatron pre-training. We hope that could refine the performance. Unfortunately, Megatron pre-training is stuck by going with kernel-free send/recv. By checking the history issuses, the situation looks like the case described in the issue [#1509](https://github.com/NVIDIA/nccl/issues/1509). 

We have been struggling with this problem, which is a show stopper for our project, for some time. It would be highly appreciated if any clue or suggestion. Many thanks.  @sjeaugey @kiskra-nvidia

**Comment by sjeaugey** [comment]
We're working on providing 0 SM communication in case where it's applicable. But as you noted, this is not an easy task, as we need to ensure deadlock-free operation, which implies being careful at which CUDA call we make and when.

Our current investigation plan is not focusing on send/recv over the network as first step though -- instead we've been looking at allgather over NVLink. But networking is an obvious next step (because it can also fully offload pure-network operations) which we plan to address in next releases. Point-to-point operations is also something we want to add after collectives. But it's likely going to be one-sided operations (e.g. ncclPut/ncclGet) rather than send/recv. Also note, all this will require buffers to be registered through ncclCommWindowRegister, so that we know remote addresses upfront and we can convert NCCL operations into network or copy engine operations directly.

Let me know if you believe that would address your use case. Thanks!

**Comment by yanminjia** [comment]
Many thanks for your response. I appreciate your time very much. It's great to know the development plan on 0 SM communication and registered buffer.

In our case, by checking the status with gdb & cuda-gdb when the program hung, I think the following currently running works on different streams could be blamed:

- default stream (forward/backward computation) - trying to launch kernel function relative to forward/backward;
- NCCL stream - a hostfunc kernel runing a trivial loop checking the flag variable which indicates the send/recv is over or not;
- copy stream - call cudaMemcpyAsync to transfer data between src/dst address and intermediate buffer.

By checking our code over and over, I don't think the deadlock was caused by some kind of programming logical mistake. I suspect it was caused by some kind of limitation of defect of CUDA scheduler in case of stream synchroniztion needed.

**Comment by sjeaugey** [comment]
CUDA does not provide forward progress guarantees, even when using non-blocking streams. Non-blocking streams doesn't mean that the stream will never block, only that it may run in parallel i.e. it will not always block.

**Comment by yanminjia** [comment]
It's really tricky. But it is a problem indeed. In fact, this problem can be reproduced by going with the native NCCL code if set NCCL_P2P_USE_CUDA_MEMCPY=1 as issue#1509 described. As you know, we have no way to know what happened by checking CUDA code which is not available. I'm not sure if I could reach out of Nvidia CUDA team for help.

**Comment by yanminjia** [comment]
It looks like cudaMemcpyAsync somehow raised the operation deadlock. But I'm not sure.

**Comment by sjeaugey** [comment]
If it's just about CE operations, you can check the NCCL code which creates a separate context for those operations in the proxy code. That usually solves the issue, although again it's not guaranteed.

**Comment by yanminjia** [comment]
As you suggested in issue#1509, we tried but failed by setting NCCL_CREATE_THREAD_CONTEXT=1 even with native NCCL code. Anyway, we will look into it along this clue. Many thanks.

**Comment by Qizhi697** [comment]
@yanminjia 
Hello, sorry to bother you. I'm currently facing the same issue and would like to ask if you have made any progress on it. I'd like to exchange ideas with you. Is it convenient for you to chat on WeChat?

**Comment by Skylion007** [comment]
Any updates on this issue btw? @yanminjia @Qizhi697 ?
